go/storage/mkvs/db/pathbadger/pathbadger.go

```diff
@@ -0,0 +1,940 @@
+// Package pathbadger provides a Badger-backed node database that uses trie paths as keys.
+package pathbadger
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/dgraph-io/badger/v4"
+
+	"github.com/oasisprotocol/oasis-core/go/common"
+	cmnBadger "github.com/oasisprotocol/oasis-core/go/common/badger"
+	"github.com/oasisprotocol/oasis-core/go/common/cbor"
+	"github.com/oasisprotocol/oasis-core/go/common/logging"
+	"github.com/oasisprotocol/oasis-core/go/storage/mkvs/db/api"
+	"github.com/oasisprotocol/oasis-core/go/storage/mkvs/node"
+	"github.com/oasisprotocol/oasis-core/go/storage/mkvs/writelog"
+)
+
+// dbVersion is the internal database version. We start with 6 to make sure this is distinct from
+// the older badger backend which uses a database version of 5.
+const dbVersion = 6
+
+// New creates a new BadgerDB-backed node database that uses trie paths as keys.
+func New(cfg *api.Config) (api.NodeDB, error) {
+	db := &badgerNodeDB{
+		logger:           logging.GetLogger("mkvs/db/pathbadger"),
+		namespace:        cfg.Namespace,
+		readOnly:         cfg.ReadOnly,
+		discardWriteLogs: cfg.DiscardWriteLogs,
+	}
+	opts := commonConfigToBadgerOptions(cfg, db.logger)
+
+	var err error
+	if db.db, err = badger.OpenManaged(opts); err != nil {
+		return nil, fmt.Errorf("mkvs/pathbadger: failed to open database: %w", err)
+	}
+
+	// Make sure that we can discard any deleted/invalid metadata.
+	db.db.SetDiscardTs(tsMetadata)
+
+	// Load database metadata.
+	if err = db.loadMetadata(); err != nil {
+		_ = db.db.Close()
+		return nil, fmt.Errorf("mkvs/pathbadger: failed to load metadata: %w", err)
+	}
+
+	// Cleanup any multipart restore remnants, since they can't be used anymore.
+	if err = db.cleanMultipartLocked(true); err != nil {
+		_ = db.db.Close()
+		return nil, fmt.Errorf("mkvs/pathbadger: failed to clean leftovers from multipart restore: %w", err)
+	}
+
+	db.gc = cmnBadger.NewGCWorker(db.logger, db.db)
+
+	return db, nil
+}
+
+type badgerNodeDB struct { // nolint: maligned
+	logger *logging.Logger
+
+	namespace common.Namespace
+
+	readOnly         bool
+	discardWriteLogs bool
+
+	multipartVersion uint64
+	multipartMeta    map[uint8]*multipartMeta
+
+	db *badger.DB
+	gc *cmnBadger.GCWorker
+
+	// metaUpdateLock must be held at any point where data at tsMetadata is read and updated. This
+	// is required because all metadata updates happen at the same timestamp and as such conflicts
+	// cannot be detected.
+	metaUpdateLock sync.Mutex
+	meta           metadata
+
+	closeOnce sync.Once
+}
+
+func (d *badgerNodeDB) loadMetadata() error {
+	tx := d.db.NewTransactionAt(tsMetadata, true)
+	defer tx.Discard()
+
+	// Ensure that no legacy metadata exists to prevent corrupting a database created using the old
+	// badger backend.
+	if _, err := tx.Get([]byte{0x04}); err != badger.ErrKeyNotFound {
```

Comment:
I would rather not import from the other database backend as this creates a strange dependency between the two.

Reply:
You are right. Probably the right approach would be to check this on higher levels, to avoid this strange dependency. But I guess we can leave it here also. Marking as resolved.

Reply:
Yeah higher layers already make sure this doesn't happen, but this is a sanity check in case someone would manually do some database directory interventions.